sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/ThroughputControlStore.java

```diff
@@ -0,0 +1,230 @@
+                    // It is caused by control store out of sync or the request has staled info.
+                    // We will handle the first scenario by creating a new container controller,
+                    // while fall back to original request Mono for the second scenario.
+                    return this.shouldRefreshContainerController(collectionLink, request)
+                        .flatMap(shouldRefresh -> {
+                            if (shouldRefresh) {
+                                containerController.close();
+                                this.containerControllerCache.refresh(collectionLink, () -> this.createAndInitContainerController(collectionLink));
+                                return this.resolveContainerController(collectionLink);
+                            }
+
+                            // The container container controller is up to date, the request has staled info, will let the request pass
+                            return Mono.just(containerController);
+                        });
+                })
+                .flatMap(containerController -> {
+                    if (containerController.canHandleRequest(request)) {
+                        return containerController.processRequest(request, originalRequestMono)
+                            .doOnError(throwable -> this.doOnError(request, containerController, throwable));
+                    } else {
+                        // still can not handle the request
+                        logger.warn(
+                            "Can not find container controller to process request {} with collectionRid {} ",
+                            request.getActivityId(),
+                            request.requestContext.resolvedCollectionRid);
+
+                        return originalRequestMono;
+                    }
+                });
+        } else {
+            return originalRequestMono;
+        }
+    }
+
+    private Mono<Boolean> shouldRefreshContainerController(String containerLink, RxDocumentServiceRequest request) {
+        // TODO: populate diagnostics
+        return this.collectionCache.resolveByNameAsync(null, containerLink, null)
+            .flatMap(documentCollection ->
+                Mono.just(StringUtils.equals(documentCollection.getResourceId(), request.requestContext.resolvedCollectionRid)));
+    }
+
+    private void doOnError(RxDocumentServiceRequest request, IThroughputController controller, Throwable throwable) {
+        checkNotNull(request, "Request can not be null");
+        checkNotNull(controller, "Container controller can not be null");
+        checkNotNull(throwable, "Exception can not be null");
+
+        CosmosException cosmosException = Utils.as(throwable, CosmosException.class);
+
+        if (cosmosException != null &&
+            (Exceptions.isNameCacheStale(cosmosException) || Exceptions.isPartitionKeyMismatchException(cosmosException))) {
+
+            controller.close();
```

Comment:
Based on my comment on `IThroughputController` extending `AutoCloseable` interface, you can decide to close this quietly, like done in other places in code. `LifeCycleUtils.closeQuietly` if you want, otherwise, this `close` may throw exceptions based on the implementation, which can mess up with the code flow.

Reply:
Discussed offline, for now will keep init() and close() method as Mono